kips/kip-2.md

```diff
@@ -0,0 +1,460 @@
+
+## Specification
+
+We introduce 2 new contracts in addition to the existing event contract:
+
+- `UserPot` - a single instance of this holds all ETH for all users across all events.
+- `ACL` - a single instance of this holds the list of Kickback admins who can upgrade the `UserPot`.
+
+When a user RSVP's to an event this is the new flow:
+
+1. User calls `register()` on the event's deployed contract with enough ETH
+2. The event's contract forwards the ETH to the global `UserPot` contract
+3. The `UserPot` adds the event address to the list of events the user is attending
+4. `UserPot` also calculates the amount of ETH withdrawable by the user (based on past event payouts owed to them) and records this number
+
+When a user wishes to withdraw their ETH:
+
+1. The user send a transaction to `UserPot` asking to withdraw
+2. `UserPot` calculates all the ETH owed to the user based on past attendance
+3. `UserPot` updates the `Storage` contract data for the user, moving events which have ended from the user's event list, and setting their withdrawable balance to 0
+4. `UserPot` sends the total ETH back to the user
+
+## Rationale
+
+The above design was arrived at after considering various alternatives as well as other desirable features that users have asked for. Here are the key points:
+
+- **Singleton user pot is necessary** - because our system redistributes ETH between users (i.e. if you don't show up your ETH gets _given_ to other users) it's necessary to have all ETH for a given event stored in one place, as it would be too expensive gas-wise and time-wise to send ETH amongst the participants of an event. This is why we can't have a `UserPot` instance per user.
+
+- **Reuse of deposit** - because we have a single user pot instance, we can easily reuse the payout from an old event as a deposit for a new event. In the sample implementation below, this is actually what happens when a user RSVP's. Thus, users do not need to withdraw their ETH and send it back in in order to attend future events.
+
+- **Easier overview of user balance** - because all ETH is in one place we can easily give the user an on-chain overview of their total deposits and payouts on Kickback.
+
+- **Upgrade-able contracts** - by making the user pot upgradeable (with admin approval) we allow for flexibility in future and for bug fixing capability.
+
+- **Configurable event incentives** - one of the nice things about this architecture is that the `UserPot` is only responsible for summing up deposits and payouts - actual calculation of what the payout should be is done by the `Event` contract. If we wished to test out new strategies (e.g. sweepstakes) or try using a single `Event` instance for multiple events we could easily do so.
+
```

Comment:
You can also add the following

- ** Potential revenue opportunity to earn interest

- With the rise of tools and projects such as [rDAI](https://rdai.money) and Dai savings rate, there will be an increasing opportunity of earning interests from the staked asset. By changing the structure to hold our customers money, it will increase the overall revenue potential when we integrate with such features in future.